File: apps/desktop/src-tauri/src/repositories/submission_repository.rs
use sqlx::{Row, SqlitePool};

pub struct VenueRow {
    pub id: String,
    pub venue_type: String,
    pub name: String,
    pub full_name: String,
    pub website: String,
    pub ccf: String,
    pub area: String,
    pub starred: bool,
    pub ei: bool,
    pub sci: bool,
    pub sci_quartile: String,
    pub deadline: Option<String>,
    pub notification_date: Option<String>,
    pub special_issue_deadline: Option<String>,
    pub special_issue_title: String,
    pub created_at: String,
}

pub struct CreateVenueInput<'a> {
    pub id: &'a str,
    pub venue_type: &'a str,
    pub name: &'a str,
    pub full_name: &'a str,
    pub website: &'a str,
    pub ccf: &'a str,
    pub area: &'a str,
    pub ei: bool,
    pub sci: bool,
    pub sci_quartile: &'a str,
    pub deadline: Option<&'a str>,
    pub notification_date: Option<&'a str>,
    pub special_issue_deadline: Option<&'a str>,
    pub special_issue_title: &'a str,
}

#[derive(Default)]
pub struct UpdateVenueInput<'a> {
    pub name: Option<&'a str>,
    pub full_name: Option<&'a str>,
    pub venue_type: Option<&'a str>,
    pub website: Option<&'a str>,
    pub ccf: Option<&'a str>,
    pub area: Option<&'a str>,
    pub ei: Option<bool>,
    pub sci: Option<bool>,
    pub sci_quartile: Option<&'a str>,
    pub deadline: Option<&'a str>,
    pub notification_date: Option<&'a str>,
    pub special_issue_deadline: Option<&'a str>,
    pub special_issue_title: Option<&'a str>,
}

pub async fn list_venues(pool: &SqlitePool) -> Result<Vec<VenueRow>, String> {
    let rows = sqlx::query(
        "SELECT id, type, name, full_name, website, ccf, area, starred, ei, sci, sci_quartile,
                deadline, notification_date, special_issue_deadline, special_issue_title, created_at
         FROM venues ORDER BY starred DESC, name ASC",
    )
    .fetch_all(pool)
    .await
    .map_err(|e| e.to_string())?;

    Ok(rows
        .into_iter()
        .map(|row| VenueRow {
            id: row.get("id"),
            venue_type: row.get("type"),
            name: row.get("name"),
            full_name: row.get("full_name"),
            website: row.get("website"),
            ccf: row.get("ccf"),
            area: row.get("area"),
            starred: row.get::<i64, _>("starred") == 1,
            ei: row.get::<i64, _>("ei") == 1,
            sci: row.get::<i64, _>("sci") == 1,
            sci_quartile: row.get("sci_quartile"),
            deadline: row.get("deadline"),
            notification_date: row.get("notification_date"),
            special_issue_deadline: row.get("special_issue_deadline"),
            special_issue_title: row.get("special_issue_title"),
            created_at: row.get("created_at"),
        })
        .collect())
}

pub async fn create_venue(pool: &SqlitePool, input: &CreateVenueInput<'_>) -> Result<(), String> {
    sqlx::query(
        "INSERT INTO venues (id, type, name, full_name, website, ccf, area, ei, sci, sci_quartile,
                             deadline, notification_date, special_issue_deadline, special_issue_title)
         VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
    )
    .bind(input.id)
    .bind(input.venue_type)
    .bind(input.name)
    .bind(input.full_name)
    .bind(input.website)
    .bind(input.ccf)
    .bind(input.area)
    .bind(input.ei as i64)
    .bind(input.sci as i64)
    .bind(input.sci_quartile)
    .bind(input.deadline)
    .bind(input.notification_date)
    .bind(input.special_issue_deadline)
    .bind(input.special_issue_title)
    .execute(pool)
    .await
    .map_err(|e| e.to_string())?;

    Ok(())
}

pub async fn update_venue(
    pool: &SqlitePool,
    id: &str,
    input: UpdateVenueInput<'_>,
) -> Result<(), String> {
    if let Some(value) = input.name {
        sqlx::query("UPDATE venues SET name = ? WHERE id = ?")
            .bind(value)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.full_name {
        sqlx::query("UPDATE venues SET full_name = ? WHERE id = ?")
            .bind(value)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.venue_type {
        sqlx::query("UPDATE venues SET type = ? WHERE id = ?")
            .bind(value)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.website {
        sqlx::query("UPDATE venues SET website = ? WHERE id = ?")
            .bind(value)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.ccf {
        sqlx::query("UPDATE venues SET ccf = ? WHERE id = ?")
            .bind(value)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.area {
        sqlx::query("UPDATE venues SET area = ? WHERE id = ?")
            .bind(value)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.ei {
        sqlx::query("UPDATE venues SET ei = ? WHERE id = ?")
            .bind(value as i64)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.sci {
        sqlx::query("UPDATE venues SET sci = ? WHERE id = ?")
            .bind(value as i64)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.sci_quartile {
        sqlx::query("UPDATE venues SET sci_quartile = ? WHERE id = ?")
            .bind(value)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.deadline {
        let normalized = if value.is_empty() { None } else { Some(value) };
        sqlx::query("UPDATE venues SET deadline = ? WHERE id = ?")
            .bind(normalized)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.notification_date {
        let normalized = if value.is_empty() { None } else { Some(value) };
        sqlx::query("UPDATE venues SET notification_date = ? WHERE id = ?")
            .bind(normalized)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.special_issue_deadline {
        let normalized = if value.is_empty() { None } else { Some(value) };
        sqlx::query("UPDATE venues SET special_issue_deadline = ? WHERE id = ?")
            .bind(normalized)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }
    if let Some(value) = input.special_issue_title {
        sqlx::query("UPDATE venues SET special_issue_title = ? WHERE id = ?")
            .bind(value)
            .bind(id)
            .execute(pool)
            .await
            .map_err(|e| e.to_string())?;
    }

    Ok(())
}

pub async fn delete_venue(pool: &SqlitePool, id: &str) -> Result<(), String> {
    sqlx::query("DELETE FROM venues WHERE id = ?")
        .bind(id)
        .execute(pool)
        .await
        .map_err(|e| e.to_string())?;

    Ok(())
}

pub async fn toggle_venue_star(pool: &SqlitePool, id: &str) -> Result<(), String> {
    sqlx::query("UPDATE venues SET starred = CASE WHEN starred = 1 THEN 0 ELSE 1 END WHERE id = ?")
        .bind(id)
        .execute(pool)
        .await
        .map_err(|e| e.to_string())?;

    Ok(())
}

File: apps/mac/Sources/XiaoYan/Database/SubmissionRepository.swift
import Foundation
import GRDB

struct SubmissionRepository {
    let dbQueue: DatabaseQueue = DatabaseManager.shared.dbQueue

    // Venues
    func listVenues(starred: Bool? = nil) throws -> [Venue] {
        try dbQueue.read { db in
            var sql = "SELECT * FROM venues"
            if let starred {
                sql += starred ? " WHERE starred = 1" : " WHERE starred = 0 OR starred IS NULL"
            }
            sql += " ORDER BY created_at DESC"
            return try Venue.fetchAll(db, sql: sql)
        }
    }

    func createVenue(_ venue: Venue) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: """
                    INSERT INTO venues (id, type, name, full_name, website, ccf, area, starred,
                        ei, sci, sci_quartile, deadline, notification_date,
                        special_issue_title, special_issue_deadline, special_issue_description)
                    VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?)
                """,
                arguments: [
                    venue.id, venue.type, venue.name, venue.fullName, venue.website,
                    venue.ccfRating, venue.area, venue.starred ?? false,
                    venue.ei, venue.sci, venue.sciQuartile,
                    venue.deadline, venue.notificationDate,
                    venue.specialIssueTitle, venue.specialIssueDeadline, venue.specialIssueDescription
                ]
            )
        }
    }

    func updateVenue(_ venue: Venue) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: """
                    UPDATE venues SET type=?, name=?, full_name=?, website=?, ccf=?, area=?,
                        starred=?, ei=?, sci=?, sci_quartile=?, deadline=?, notification_date=?,
                        special_issue_title=?, special_issue_deadline=?, special_issue_description=?
                    WHERE id=?
                """,
                arguments: [
                    venue.type, venue.name, venue.fullName, venue.website,
                    venue.ccfRating, venue.area, venue.starred ?? false,
                    venue.ei, venue.sci, venue.sciQuartile,
                    venue.deadline, venue.notificationDate,
                    venue.specialIssueTitle, venue.specialIssueDeadline, venue.specialIssueDescription,
                    venue.id
                ]
            )
        }
    }

    func deleteVenue(id: String) throws {
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM venues WHERE id = ?", arguments: [id])
        }
    }

    func toggleVenueStar(id: String) throws {
        try dbQueue.write { db in
            try db.execute(sql: "UPDATE venues SET starred = NOT starred WHERE id = ?", arguments: [id])
        }
    }

    // Submissions
    func listSubmissions() throws -> [Submission] {
        try dbQueue.read { db in
            try Submission.fetchAll(db, sql: "SELECT * FROM submissions ORDER BY created_at DESC")
        }
    }

    func createSubmission(_ submission: Submission) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: "INSERT INTO submissions (id, title, venue_name, venue_type, status, deadline) VALUES (?,?,?,?,?,?)",
                arguments: [submission.id, submission.title, submission.venueName, submission.venueType, submission.status?.rawValue, submission.deadline]
            )
        }
    }

    func updateSubmission(_ submission: Submission) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: "UPDATE submissions SET title=?, venue_name=?, venue_type=?, status=?, deadline=?, submitted_at=? WHERE id=?",
                arguments: [
                    submission.title, submission.venueName, submission.venueType,
                    submission.status?.rawValue, submission.deadline, submission.submittedAt, submission.id
                ]
            )
        }
    }

    func deleteSubmission(id: String) throws {
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM submissions WHERE id = ?", arguments: [id])
        }
    }

    // Review
    func insertReviewRound(_ round: ReviewRound) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: "INSERT INTO review_rounds (id, submission_id, round, verdict, received_at) VALUES (?,?,?,?,?)",
                arguments: [round.id, round.submissionId, round.round, round.verdict, round.receivedAt]
            )
        }
    }

    func upsertReviewRound(_ round: ReviewRound) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: """
                    INSERT INTO review_rounds (id, submission_id, round, verdict, received_at)
                    VALUES (?,?,?,?,?)
                    ON CONFLICT(id) DO UPDATE SET
                      verdict = excluded.verdict,
                      received_at = excluded.received_at
                """,
                arguments: [round.id, round.submissionId, round.round, round.verdict, round.receivedAt]
            )
        }
    }

    func insertReviewComment(_ comment: ReviewComment) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: "INSERT INTO review_comments (id, submission_id, round, reviewer, content, response, resolved, tags, verdict) VALUES (?,?,?,?,?,?,?,?,?)",
                arguments: [
                    comment.id, comment.submissionId, comment.round, comment.reviewer,
                    comment.content, comment.response, comment.resolved ?? false,
                    comment.tags?.jsonString, comment.verdict
                ]
            )
        }
    }

    func updateReviewComment(_ comment: ReviewComment) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: "UPDATE review_comments SET reviewer=?, content=?, response=?, resolved=?, tags=?, verdict=? WHERE id=?",
                arguments: [
                    comment.reviewer, comment.content, comment.response, comment.resolved ?? false,
                    comment.tags?.jsonString, comment.verdict, comment.id
                ]
            )
        }
    }

    // Versions
    func listVersions(submissionId: String) throws -> [PaperVersion] {
        try dbQueue.read { db in
            try PaperVersion.fetchAll(db, sql: "SELECT * FROM paper_versions WHERE submission_id = ? ORDER BY created_at DESC", arguments: [submissionId])
        }
    }

    func insertVersion(_ version: PaperVersion) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: "INSERT INTO paper_versions (id, submission_id, tag, label, stage, content, notes, file_path, file_name) VALUES (?,?,?,?,?,?,?,?,?)",
                arguments: [
                    version.id, version.submissionId, version.tag, version.label,
                    version.stage, version.content, version.notes, version.filePath, version.fileName
                ]
            )
        }
    }

    func deleteVersion(id: String) throws {
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM paper_versions WHERE id = ?", arguments: [id])
        }
    }

    func updateVersion(_ version: PaperVersion) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: """
                    UPDATE paper_versions
                    SET tag = ?, label = ?, stage = ?, content = ?, notes = ?,
                        file_path = ?, file_name = ?, updated_at = CURRENT_TIMESTAMP
                    WHERE id = ?
                """,
                arguments: [
                    version.tag, version.label, version.stage, version.content, version.notes,
                    version.filePath, version.fileName, version.id
                ]
            )
        }
    }

    // Review Rounds
    func listReviewRounds(submissionId: String) throws -> [ReviewRound] {
        try dbQueue.read { db in
            try ReviewRound.fetchAll(db, sql: "SELECT * FROM review_rounds WHERE submission_id = ? ORDER BY round ASC", arguments: [submissionId])
        }
    }

    func listReviewComments(submissionId: String, round: Int) throws -> [ReviewComment] {
        try dbQueue.read { db in
            try ReviewComment.fetchAll(db, sql: "SELECT * FROM review_comments WHERE submission_id = ? AND round = ? ORDER BY created_at ASC", arguments: [submissionId, round])
        }
    }

    func deleteReviewRound(id: String) throws {
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM review_rounds WHERE id = ?", arguments: [id])
            try db.execute(sql: "DELETE FROM review_comments WHERE id IN (SELECT id FROM review_comments WHERE round IN (SELECT round FROM review_rounds WHERE id = ?))", arguments: [id])
        }
    }

    // Stats
    struct SubmissionStats {
        let active: Int
        let pendingReviews: Int
        let upcomingDdls: [(name: String, deadline: Date)]
    }

    func stats() throws -> SubmissionStats {
        try dbQueue.read { db in
            let active = try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM submissions WHERE status IN ('writing', 'submitted', 'reviewing')") ?? 0
            let pendingReviews = try Int.fetchOne(db, sql: "SELECT COUNT(*) FROM review_comments WHERE resolved = 0 OR resolved IS NULL") ?? 0
            let rows = try Row.fetchAll(db, sql: "SELECT venue_name, deadline FROM submissions WHERE deadline IS NOT NULL AND status IN ('writing', 'submitted', 'reviewing') ORDER BY deadline ASC LIMIT 5")
            let upcomingDdls = rows.compactMap { row -> (name: String, deadline: Date)? in
                guard let deadline = row["deadline"] as Date? else { return nil }
                return (name: row["venue_name"] as? String ?? "未命名", deadline: deadline)
            }
            return SubmissionStats(active: active, pendingReviews: pendingReviews, upcomingDdls: upcomingDdls)
        }
    }

    // Checklist
    func upsertChecklistItem(_ item: SubmissionChecklistItem) throws {
        try dbQueue.write { db in
            try db.execute(
                sql: "INSERT INTO submission_checklist (id, submission_id, label, checked, category, sort_order) VALUES (?,?,?,?,?,?) ON CONFLICT(id) DO UPDATE SET label=excluded.label, checked=excluded.checked, category=excluded.category, sort_order=excluded.sort_order",
                arguments: [item.id, item.submissionId, item.label, item.checked ?? false, item.category, item.sortOrder ?? 0]
            )
        }
    }

    func deleteChecklistItem(id: String) throws {
        try dbQueue.write { db in
            try db.execute(sql: "DELETE FROM submission_checklist WHERE id = ?", arguments: [id])
        }
    }
}

private extension Encodable {
    var jsonString: String {
        (try? JSONEncoder().encode(self)).flatMap { String(data: $0, encoding: .utf8) } ?? "[]"
    }
}
